src/uart.rs

```diff
@@ -0,0 +1,375 @@
+    /// uart.write(b"Hello from UART");
+    /// ```
+    pub fn new_with_static_cb<F>(
+        index: usize,
+        baud: u32,
+        user_callback: &'static mut F,
+    ) -> Result<Self, UartDeviceError>
+    where
+        F: FnMut(u8) + Send + 'static,
+    {
+        unsafe { Self::construct_uart(index, baud, user_callback) }
+    }
+
+    /// Sets the mode according to the given parameters
+    /// Should the parameters be invalid, the function returns a Err<UartDeviceStatus::UnsupportedConfig>
+    /// # Arguments
+    /// * `data_bits` - Number of data bits in a UART frame
+    /// * `parity` - Parity mode
+    /// * `stop_bits` - Number of stop bits in a UART frame
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::{DataBits, Parity, StopBits, UartDevice};
+    /// let mut uart = UartDevice::new_without_rx(0, 115200)
+    /// .unwrap_or_else(|e| panic!("Error initializing UART: {e:?}"));   
+    /// uart.set_mode(DataBits::Eight, Parity::None, StopBits::One)   
+    /// .unwrap_or_else(|e| panic!("Error setting UART mode: {e:?}"));
+    /// ```
+    #[cfg(riot_module_periph_uart_modecfg)]
+    pub fn set_mode(
+        &mut self,
+        data_bits: DataBits,
+        parity: Parity,
+        stop_bits: StopBits,
+    ) -> Result<(), UartDeviceError> {
+        unsafe {
+            match UartDeviceError::from_c(uart_mode(
+                self.dev,
+                data_bits.to_c(),
+                parity.to_c(),
+                stop_bits.to_c(),
+            )) {
+                UartDeviceError::Success => Ok(()),
+                status => Err(status),
+            }
+        }
+    }
+
+    /// Transmits the given data via the `UART`-device
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::UartDevice;
+    /// let mut uart = UartDevice::new_without_rx(0, 115200)
+    ///    .unwrap_or_else(|e| panic!("Error initializing UART: {e:?}"));
+    /// uart.write(b"Hello from UART\n");
+    /// ```
+    pub fn write(&mut self, data: &[u8]) {
+        unsafe {
+            uart_write(self.dev, data.as_ptr(), data.len() as size_t);
+        }
+    }
+
+    /// Turns on the power from the `UART-Device`
+    pub fn power_on(&mut self) {
+        unsafe { uart_poweron(self.dev) };
+    }
+
+    /// Turns off the power from the `UART-Device`
+    pub fn power_off(&mut self) {
+        unsafe { uart_poweroff(self.dev) };
+    }
+
+    /// This function normally does not need to be called. But in some case, the pins on the `UART`
+    /// might be shared with some other functionality (like `GPIO`). In this case, it is necessary
+    /// to give the user the possibility to init the pins again.
+    #[cfg(riot_module_periph_uart_reconfigure)]
+    pub unsafe fn init_pins(&mut self) {
+        uart_init_pins(self.dev);
+    }
+
+    /// Change the pins back to plain GPIO functionality
+    #[cfg(riot_module_periph_uart_reconfigure)]
+    pub unsafe fn deinit_pins(&mut self) {
```

Comment:
When is this safe to call?

Reply:
Hmm, I'm not so sure about this. Is it an option to consume the object here via taking `self`, so that the is destructed, and the user can use the pins as normal GPIO?

Reply:
If the object is consumed, it can't be reinited with the same settings later; it's an option, but I don't think that's the intention of deinit_pins. That seems to be more of a "let's pause this for a moment, do funky stuff with the UART, and return to using it later".

If we're not sure (I am not either), maybe let's remove that API for now.

If we keep this and the other removals in a separate commit when squashing, it'll be easy to selectively revert it.